pkg/kubelet/cpumanager/cpu_manager.go

```diff
@@ -0,0 +1,363 @@
+)
+
+type kletGetter interface {
+	GetPods() []*v1.Pod
+	GetCachedMachineInfo() (*cadvisorapi.MachineInfo, error)
+}
+
+type runtimeService interface {
+	UpdateContainerResources(id string, resources *runtimeapi.LinuxContainerResources) error
+}
+
+type policyName string
+
+// Manager interface provides methods for kubelet to manage pod cpus
+type Manager interface {
+	Start()
+
+	// RegisterContainer registers a container with the cpuset manager
+	// resulting in synchronous setting on the container cpuset.  This
+	// is called after RegisterContainer(), which creates the containerID, and
+	// before StartContainer(), such that the cpuset is configured before
+	// the container starts
+	RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error
+
+	// UnregisterContainer is called near UnregisterContainer() so that the manager
+	// stops trying to update that container in the reconcilation loop and
+	// any CPUs dedicated to the container are freed to the shared pool.
+	UnregisterContainer(containerID string) error
+
+	State() state.Reader
+
+	IsUnderCPUPressure() bool
+
+	lifecycle.PodAdmitHandler
+}
+
+type manager struct {
+	sync.Mutex
+	policy Policy
+	state  state.State
+
+	// containerRuntime is the container runtime service interface needed
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+
+	// kletGetter provides a method for listing all the pods on the node
+	// so all the containers can be updated in the reconciliation loop.
+	kletGetter kletGetter
+
+	// podStatusProvider provides a method for obtaining pod statuses
+	// and the containerID of their containers
+	podStatusProvider status.PodStatusProvider
+
+	// getPodsFunc gets list of active pods to consider for eviction when
+	// under pressure.
+	getPodsFunc eviction.ActivePodsFunc
+
+	// killPodFunc kills a pod selected for eviction when under pressure.
+	killPodFunc eviction.KillPodFunc
+
+	// recorder records eviction events.
+	recorder record.EventRecorder
+}
+
+var _ Manager = &manager{}
+
+// NewManager creates new cpu manager based on provided policy
+func NewManager(
+	cpuPolicyName string,
+	cr runtimeService,
+	kletGetter kletGetter,
+	statusProvider status.PodStatusProvider,
+	capacityProvider eviction.CapacityProvider,
+	getPodsFunc eviction.ActivePodsFunc,
+	killPodFunc eviction.KillPodFunc,
+	recorder record.EventRecorder,
+) (Manager, lifecycle.PodAdmitHandler, error) {
+	var policy Policy
+
+	switch policyName(cpuPolicyName) {
+	case PolicyNone:
+		policy = NewNonePolicy()
+	case PolicyStatic:
+		machineInfo, err := kletGetter.GetCachedMachineInfo()
+		if err != nil {
+			return nil, nil, err
+		}
+		topo, err := topology.Discover(machineInfo)
+		if err != nil {
+			return nil, nil, err
+		}
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+
+		resources := capacityProvider.GetNodeAllocatableReservation()
+		cpuResource, ok := resources[v1.ResourceCPU]
```

Comment:
The situation you described to me sounds like an operator error. 
To avoid having users be baffled,
1. Cluster admins can be instructed to reserve CPU for system daemons and,
2. Kubelet can reject pods during admission that ask for more than available "exclusive" CPUs.
The approach of using node pressure condition for representing schedulable capacity seems like an anti-pattern. Thus far pressure conditions have been about usage.

We may help users by having kubelet require that `kubereserved.cpu + systemreserved.cpu > 0 millicpus` when static CPU pinning policy is enabled. 

I'm suggesting identifying ways that avoid the problem you mention without requiring much changes to the system.

> matches the wider discussion that occurred during proposal review time.

I apologize, I did recall the discussion around requiring users to set non zero cpu allocatable, but I did not recall requiring a new node condition.
This feature is non trivial. Let's not rush to adding an API just because of release pressure.
Adding a node condition is easy, but removing one is hard.